Linearize bilinear terms of lower dual*primal in upper objective

NLaws opened the pull request:
## Summary
When the upper level objective contains bilinear products of lower dual and primal variables they can be linearized under certain conditions. The changes in this PR checks those conditions and linearizes the biinear terms when possible. The mathematical details can be found in:
http://www.optimization-online.org/DB_HTML/2021/08/8561.html

## Changes
- add an optional flag to `BilevelModel` called `linearize_bilinear_upper_terms`, which is a boolean that defaults to `false`.
- add automated test for the new capability using the existing `jump_conejo2016` test.
- address https://github.com/joaquimg/BilevelJuMP.jl/issues/158 by adding support for constructing `DualOf` vectors and DenseAxisArrays of constraint references

## Additional information
There are some additional, manual tests in a new file for the linearization process that I am keeping in the code for now. They will be useful to adapt and add to the automated tests as this new capability is expanded.
For example, currently the linearization process requires that the lower level model be provided by the user in its standard form. However, I believe that we could do the conversion of the lower level to standard form for users (and I have a start on it in the new functions). I will need help with that because I'm not sure how to handle the dualizing and all the maps if we convert the lower level to standard form before building the single level model. (I think that we will need a map from/to:
single level model to/from the dualized standard form lower model to/from the general form primal lower model).

NLaws committed:
add support functions for linearizing bilinear terms

NLaws committed:
first draft of bilinear linearization

NLaws committed:
bilinear linearization: handle different lower level obj types

and starting to handle case when set AB is empty (no shared lower primal variables in bilinear upper and lower objectives). checking with jump_conejo2016 but linearized results do not match the bilinear problem results yet

NLaws committed:
add methods to create standard form of lower level

NLaws committed:
wip on testing bilinear linearization

NLaws committed:
more work in linearization of bilinear terms

NLaws committed:
handle undefined bounds in standard form of lower level

NLaws committed:
add temporary dev notes

NLaws committed:
check conditions for bilinear linearization

and move more methods to src/bilinear_linearization.jl

NLaws committed:
minor formatting/house cleaning

NLaws committed:
address some TODO's in bilnear linearization

still not working quite for Conejo, seem to be getting wrong dual variable indices, also will likely need to Dualize the standard form problem for linearization to work

NLaws committed:
bilinear linearization tests passing w/std form LL

tests not automated yet

NLaws committed:
minor formatting

NLaws committed:
add warnings when cannot linearize

and fill out some doc strings

NLaws committed:
mv all of linearization functions into moi.jl

and clean up (manual) tests

NLaws committed:
add jump_conejo2016_linearize to test suite

NLaws committed:
add bilinear linearization paper draft reference

NLaws committed:
rm `BilevelJuMP.`calls used for debugging

NLaws committed:
add test/Project.toml

try to get SCIP working on Actions

NLaws commented:
trying to close and reopen to kick off Actions tests

NLaws committed:
try to run tests on linux/windows/mac

NLaws commented:
@joaquimg I tried to get the tests running on actions by updating the testing framework to use test/Project.toml but it is still expecting the old format:
```
ERROR: Compat `Cbc` not listed in `deps` or `extras` section.
```
I can't find any info online on how to upgrade a package to the new Pkg testing framework. At first, Actions was failing because it could not add SCIP. Some guidance on how to get the test suite running would be greatly appreciated.

NLaws committed:
Revert "add test/Project.toml"

This reverts commit f5ac4bc05ac826380081459f3410e1191c67adfe.

NLaws committed:
add support for @variable(model, dual_var_name, DualOf(vector_of_conrefs))

hopefully addresses https://github.com/joaquimg/BilevelJuMP.jl/issues/158

NLaws committed:
handle DenseAxisArray of conrefs in @variable with DualOf

for example:
```julia

m = BilevelModel();
T = 100;
nodes = [:a, :b];
@variable(Lower(m), y1[nodes, 1:T] >= 0);
@variable(Lower(m), y2[nodes, 1:T] >= 0);
@constraint(Lower(m), b[n in nodes, t in 1:T], y1[n,t] + y2[n,t] == 100);

@variable(Upper(m), lambda, DualOf(b))

2-dimensional DenseAxisArray{BilevelJuMP.BilevelVariableRef,2,...} with index sets:
    Dimension 1, [:a, :b]
    Dimension 2, Base.OneTo(100)
And data, a 2×100 Matrix{BilevelJuMP.BilevelVariableRef}:
 lambda[1]  lambda[3]  lambda[5]  lambda[7]  lambda[9]   …  lambda[193]  lambda[195]  lambda[197]  lambda[199]
 lambda[2]  lambda[4]  lambda[6]  lambda[8]  lambda[10]     lambda[194]  lambda[196]  lambda[198]  lambda[200]
```

NLaws committed:
switch @warn's to @debug's in bilinear_linearization

NLaws committed:
handle underdetermined system in bilinear linearization

NLaws committed:
fix true/false for redundant values in recursive_col_search

NLaws committed:
return redundant_vals bool from find_connected_rows_cols

NLaws committed:
handle underdetermined system better in bilinear_linearization

NLaws committed:
address TODO for quadratic term linearization in moi.jl

NLaws committed:
handle separable lower level problems in bilinear linearization

NLaws committed:
handle DualOf not used in UL in bilinear linearization

NLaws committed:
fix mistake in moi.jl

joaquimg commented:
Awsome! I will take a look in the next days

NLaws committed:
fix bug for single block V in bilinear linearization

NLaws committed:
pre-allocate for efficiency in bilinear_linearization.jl

NLaws committed:
make Vblocks more efficient in bilinear linearization

NLaws commented:
@joaquimg I am finding that for very large problems any loops over all variables or constraints takes a large amount of time. For example, the [build_bounds!](https://github.com/joaquimg/BilevelJuMP.jl/blob/059543614d1e8b6f59f285adeeb3ba8a43697242/src/jump.jl#L645) method loops over all variables and constraints when using `FortunyAmatMcCarlMode`. If there are ~20 lines of code that each take 0.001 seconds, then a problem with 1 million variables will take more than 5 hours to loop over all the variables. Is it possible to parallelize the `for` loops over variables and constraints? If you have suggestions for how to do that I can work on it in this PR.

Commit:
Merge remote-tracking branch 'upstream/master' into bilinear

Commit:
Update moi.jl

Commit:
typo fix in moi.jl

Commit:
checkout upstream/master ci.yml

Commit:
Pkg.rm MibS_jll

Commit:
address some TODO's

NLaws commented:
@joaquimg Can you please review this next chance you get? 
(The paper has been accepted to IEEE Transactions on Power Systems and will be published soon!)

joaquimg commented:
I am trying to get to it. I had to stabilize some pieces of the code.
It is probably worth integrating master here so you can get xpress on CI.

joaquimg commented:
Great news on the acceptance! They are usually very fast at posting a preview. Link here whenever there is something!

joaquimg commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                    check_upper_objective_for_bilinear_linearization(upper, upper_to_lower_var_indices, upper_var_to_lower_ctr)
+
+                if isempty(A_N)
+                    @warn("No bilinear products of lower level dual and primal variables found in upper level objective. Skipping linearization process.")
```
this is only good for testing

joaquimg replied:
these ones that are only for tests should be `@debug`

NLaws replied:
I actually meant all @warn's for users. In the case that a user provides `linearize_bilinear_upper_terms=true` but the model does not contain valid bilinear products, then this warning will notify the user that their model is not  a candidate for linearization.

joaquimg replied:
I mean, if linearization is active but there is nothing to do, just use `@debug`, but if it failed to linearized due to some condition not being met, then use `@warn`

NLaws replied:
got it. change to debug in 0d7c8a0

joaquimg commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                end
+            end
+        else
+            @warn("Upper objective must be quadratic and there must lower dual variables in the upper model to linearize bilinear terms. Skipping linearization process.")
```
also only for tests right?

NLaws replied:
Meant for users that provide `linearize_bilinear_upper_terms=true` but do not have a model that is appropriate for linearization.

joaquimg replied:
I mean, if linearization is active but there is nothing to do, just use @debug, but if it failed to linearized due to some condition not being met, then use @warn

NLaws replied:
Got it, addressed in b02f8bc

joaquimg commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                        @warn("Unable to linearize bilinear terms due to underdertermined system of equations. Skipping linearization process.")
+                    end
+                else
+                    @warn("Required conditions for linearization not met. Skipping linearization process.")
```
can you say which? referencing the paper seems fine.

NLaws replied:
Yes the @debug statements tell users which conditions were not meant. Would you prefer that the unmet conditions are collected into this @warn statement?

joaquimg replied:
in this case I meant, lets add more details about what condition is failing to be met, keep as warning.

NLaws replied:
addressed in b02f8bc

joaquimg commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+            end
+
+            # LINEARIZATION
+            # TODO switch signs with MAX sense?
```
maybe for now add an assert that we have a min sense

NLaws replied:
done in 3f96626ceef9713d84f5585e3a2d9621d843d8bb

joaquimg commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                cvb = collect(values(bilinear_upper_quad_term_to_m_quad_term))
+                new_objective = deepcopy(m_objective)
+                if Set(cvb) == Set(quadratic_terms)
+                    @info("Replacing bilinear lower dual * lower primal terms in upper objective with linear terms.")
```
`@debug`

NLaws replied:
done in bc4539998c421fec88da7461efe8ebcdbdb6d0b9

joaquimg commented on src/bilinear_linearization.jl:
```diff
@@ -0,0 +1,811 @@
+#=
```
we really need small test cases for all corner cases.

NLaws replied:
I am adding more tests and trying to meet the codecov requirements

joaquimg reviewed (changes requested):
This is a great contribution and I am keen on merging when (minimally) ready. Thank you so much for giving it a try.

1 - I would be fine merging a version that throws errors/warnings if strict conditions is not satisfied.
However, we really need a set of small tests covering all corner cases, so that I can improve performance and organization later on without silently breaking things.

2 - if you can put renames and vector forms DualOf in a separate PR we can merge them much earlier (which would already be good contributions)

3 - we need to rebase this by integrating current master where Ci is smoother.

4 - I will think more about the conversion to standard form. My current take is: if linearization is active:
before doing anything in build_bilevel
a - check if LL is LP and there are products that have a chance to be linearized (this is basically checking ocntraint type and loop in the upper objective) if all goo continue to b, otherwise skip the linearization.
b - create an MOI model for the standard lower level (keep important maps from the original problem to this (this can be done with a copy_to moving data from the original data to a MOI model that only supported standard form so that MOI bridges will do all conversions for you) - keep maps from this standardized model to the original LL
c - extract matrices using MatrixOfConstraints using another copy_to (we can follow DiffOpt.jl example)
d - check all other conditions from matrices in step c, it fails forget linearization and proceeded as usual, if good continue to e
e - use the model from b as if you were using the original LL - we will have to be careful with maps so that solution can be recovered, but this is doable.

5 - As for build bounds, there are probably lots of type instability. Can you create an issue with a MWE that takes like 5 seconds (that should be zero). Smallest problem that you can see some meaningful slowness...

6 - I tried to respond to all your previous comments in this single shot, please let me know if I missed any.

And again, awesome work, lets try to merge this asap.

NLaws commented:
Thank you @joaquimg! I will respond to your comments that I have answers for now; and I will work on MWEs/more tests soon.

NLaws commented on test/bilinear_linearization.jl:
```diff
@@ -0,0 +1,339 @@
+# using Ipopt
+# using Cbc
+# using CPLEX
+using Gurobi
```
I do have access to Xpress and I have merged the latest master. The Xpress/CI tests are passing for this PR (but the codecov tests are not). What do you recommend to address the codecov tests? (I am unfamiliar with them, but it looks like I need more doc strings and/or tests?)

joaquimg replied:
for codecov we basically need more small tests testing different bits of the overall code. It will tell you which pieces of code are never reached, so that you can add testes that reach those lines.
Focus on your added functions.

joaquimg replied:
dont worry about docstrings now

NLaws commented on test/bilinear_linearization.jl:
```diff
@@ -0,0 +1,339 @@
+ is not including the complementary constraints for lower bounds of zero 
+(but is including the dual variables for lower bounds of zero in the dual constraints, which results in the odd solution of UL cost 0, LL cost 3)
+
+Bug in v 0.3.5 of Dualization.jl see https://github.com/jump-dev/Dualization.jl/issues/123 and address by checking out master branch of BilevelJuMP (Done)
```
done. I will remove these notes.

Commit:
Update bilinear_linearization.jl

Commit:
change info to debug in moi.jl

Commit:
add asserts for MIN_SENSE in bilinear linearization

Commit:
move main linearization into function (from moi.jl)

NLaws commented on src/moi.jl:
```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+    =#
+    # TODO implement for bilinear products in constraints (only checking upper objective so far)
+
+    if linearize_bilinear_upper_terms 
```
done in 5ba028ee13fac95bce8350b5656b44fd625eb912

NLaws committed:
change warn to debut in bilinear_linearization

when nothing to be linearized

NLaws commented on src/jump_constraints.jl:
```diff
@@ -169,16 +169,60 @@ end
     ci::BilevelConstraintRef
 end
+
+function DualOf(cis::AbstractVector{<:BilevelConstraintRef})
+    return [DualOf(ci) for ci in cis]
+end
+
+function DualOf(cis::JuMP.Containers.DenseAxisArray{<:BilevelJuMP.BilevelConstraintRef})
```
> 1 - Maybe a single method defined for `AbstractArray` calling `similar` to initialize and `eachindex` to fill would work for all 3 main cases: `Array`, `DenseAxisArray`, `SparseAxisArray`

I'm not sure what you mean. Can you please implement or provide an example?

NLaws replied:
I used your new unit test for `DualOf` a vector constraint

NLaws committed:
change bilinear linearization condition debugs to warns

and expand their explanations

NLaws committed:
change warn to debug when No bilinear products of lower level dual and primal variables found in upper level objective

NLaws committed:
rm debugging code for jump_conejo2016_linearization test

NLaws committed:
add test_recursive_col_search

(figuring out how codecov works)

NLaws committed:
add test_find_connected_rows_cols

NLaws committed:
add bilinear linearization test for lower obj. with upper*lower var.

NLaws committed:
fix test_bilinear_linearization

(testing locally with Gurobi)

NLaws committed:
comment out failing test jump_eq_price

- (testing locally with Gurobi)
- this test does not run in CI because on SCIP is tested.

NLaws committed:
rm debug tests in test/bilinear_linearization.jl

and rename one test

NLaws committed:
mv some code from build_bilevel into function to facilitate unit tests

NLaws committed:
break out code into more sub functions (for unit testing)

NLaws committed:
add tests for failing_conditions_non_empty_AB_N

joaquimg commented:
That is great!

NLaws committed:
add bilinear linearization test for condition 2'

NLaws committed:
add bilinear linearization test for condition 3

and handle repeated lower dual variables in bilinear terms

NLaws committed:
typo fix in bilinear_linearization

NLaws committed:
add test for condition 4 in bilinear linearization

NLaws committed:
add test for condition 5 in bilinear linearization

NLaws commented:
@joaquimg how does this codecov thing work? I added an explicit test for one of my new functions https://github.com/joaquimg/BilevelJuMP.jl/pull/157/commits/d1e893e828bee3719a90770c85afd5f1edb9b2e1 but the [codecov still shows that this code is untested]( https://app.codecov.io/gh/joaquimg/BilevelJuMP.jl/compare/157/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr%20comments&utm_term=Joaquim%20Dias%20Garcia#D4L761)?

NLaws committed:
Update bilinear_linearization.jl

NLaws committed:
have to have function calls outside of @test for codecov to consider them tested?

joaquimg commented:
codecov will literally check which lines were touched during tests.
it some rare cases, it might count something wrong due inlining.

joaquimg commented:
Now `master` branch is in JuMP 1.0!
So you can use update versions of anything if you need.

joaquimg committed:
rename mibs file

NLaws committed:
merge master

NLaws committed:
MOI.SingleVariable -> MOI.VariableIndex

and typo fix

NLaws committed:
MOI.ListOfConstraints -> ListOfConstraintTypesPresent

NLaws committed:
MOI.variable_index_1/2 -> variable_1/2

NLaws committed:
MOI.ScalarAffineTerm.variable_index -> variable

NLaws committed:
another MOI v1 fix

NLaws committed:
Merge branch 'master' into bilinear

NLaws committed:
fix find_blocks method for UnderDetermined UV

NLaws committed:
add more test/bilinear_linearization.jl

NLaws committed:
Merge branch 'master' into bilinear

NLaws committed:
test UnderDetermined system

NLaws committed:
Update jump_unit.jl

NLaws committed:
Update runtests.jl

NLaws commented on test/bilinear_linearization.jl:
```diff
@@ -0,0 +1,339 @@
+
```
done

NLaws commented on test/bilinear_linearization.jl:
```diff
@@ -0,0 +1,339 @@
+@constraint(model, [y[2], mu_2_lo] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [y[3], mu_3_lo] in MOI.SOS1([1.0, 2.0]))
+@constraint(model, [s, mu_s_lo] in MOI.SOS1([1.0, 2.0]))
+# seems like maybe the mu variables are not constrained to be positive in BilevelJuMP ? have to add these constraints manually?
```
I don't think this was the issue (hence the ?, it was more a note to self).

joaquimg commented on src/jump_constraints.jl:
```diff
@@ -191,29 +191,60 @@ end
-        "@variable(Upper(m), variable_type = DualOf(my_constraint_vector[t]))"
-    )
+
+function DualOf(cis::AbstractVector{<:BilevelConstraintRef})
+    return [DualOf(ci) for ci in cis]
+end
```
The error message was detailing work arounds that are more "JuMP compliant"

NLaws replied:
Sorry, I'm not understanding. What does the form of the right hand side of a constraint have to do with the `DualOf` function?

NLaws replied:
Are you saying that [this](https://github.com/NLaws/CoDER_examples/blob/4e9667b5fce2bdf9cad0664df3207fab5526bc91/planning_bileveljump.jl#L63) is wrong?
```julia
# LL load balance with PV (lambda)
@constraint(Lower(model), loadbalance[n in LLnodes_withPV, t in 1:T], 
    -ye[n,t] + yi[n,t] + ypvprod[n,t] - LDFinputs.Pload[n][t] == 0
);

# duals of LL load balances
@variable(Upper(model), 0 <= lambda <= M, DualOf(loadbalance));
```

NLaws committed:
add test_get_coef_matrix_and_rhs_vec

NLaws committed:
rm dev notes

NLaws committed:
update test/bilinear_linearization.jl

NLaws committed:
update test/bilinear_linearization.jl

NLaws commented:
@joaquimg any idea why non-linear tests would start failing?
```julia
Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2548
  Expression: ≈(objective_value(model), 12, atol = 0.1)
   Evaluated: 0.0 ≈ 12 (atol=0.1)

Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2549
  Expression: ≈(value(x), 6, atol = 0.001)
   Evaluated: 1.9596051811680095 ≈ 6 (atol=0.001)

Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2551
  Expression: ≈(value(y[1]) - value(y[2]), 2, atol = 0.001)
   Evaluated: 1.5098463839385659 ≈ 2 (atol=0.001)
```

NLaws committed:
comment out test jump_conic02

s.t. I can see codecov results

NLaws committed:
update test/bilinear_linearization.jl

NLaws committed:
update test/bilinear_linearization.jl

NLaws committed:
update test/bilinear_linearization.jl

NLaws committed:
update test/bilinear_linearization.jl

NLaws commented:
> @joaquimg any idea why non-linear tests would start failing?

The failures seem random. Now they are passing (or maybe they are not always running?)

NLaws committed:
Revert "handle DenseAxisArray of conrefs in @variable with DualOf"

This reverts commit 92055a40d246a92e5e9f323736a7130b8c18ffd4.

NLaws committed:
Revert "add support for @variable(model, dual_var_name, DualOf(vector_of_conrefs))"

This reverts commit db59f6ac62240fed02abe267678b5e11bdee981e.

NLaws committed:
restore master branch DualOf

NLaws committed:
update IEEE paper link

NLaws committed:
get more coverage w/o solver

NLaws commented:
@joaquimg I'm not sure what else I can do about the codecov. From what I can tell it is now dinging me for coverage reductions for lines that I did not add. For example, it says that there is a -57% coverage reduction for intervals.jl and a -31% reduction in moi.jl.

joaquimg commented:
> @joaquimg any idea why non-linear tests would start failing?
> 
> ```julia
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2548
> 
>   Expression: ≈(objective_value(model), 12, atol = 0.1)
> 
>    Evaluated: 0.0 ≈ 12 (atol=0.1)
> 
> 
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2549
> 
>   Expression: ≈(value(x), 6, atol = 0.001)
> 
>    Evaluated: 1.9596051811680095 ≈ 6 (atol=0.001)
> 
> 
> 
> Bilevel Conic JuMP NLP: Test Failed at /Users/nlaws/Projects/BilevelJuMP.jl/test/jump.jl:2551
> 
>   Expression: ≈(value(y[1]) - value(y[2]), 2, atol = 0.001)
> 
>    Evaluated: 1.5098463839385659 ≈ 2 (atol=0.001)
> 
> ```

Ipopt can converge to bad points.
New versions, change or orders of things in the models, different OS, might lead to that.

joaquimg commented:
> @joaquimg I'm not sure what else I can do about the codecov. From what I can tell it is now dinging me for coverage reductions for lines that I did not add. For example, it says that there is a -57% coverage reduction for intervals.jl and a -31% reduction in moi.jl.

I will look carefully through coverage later. Focus your coverage work in you new code. Try keep adding tests that result in different paths your code can go. That will be enough for me.
I can do the polishing.

NLaws commented:
@joaquimg I think that we are good with test coverage of my contributions: ~400 out of ~1600 of the new lines are tests and src/bilinear_linearization.jl has ~95% coverage. Please let me know if there is anything specific that you would like me to address.

joaquimg commented:
Awesome. I will do another pass.

NLaws commented:
@joaquimg are you going to merge this PR? If not I will close it.

NLaws committed:
Merge remote-tracking branch 'upstream/master' into bilinear

joaquimg commented:
I love this PR and I intend to move forward with this, yes.
Sorry for the delay.
My biggest issue is that it does not apply in all cases, which is totally fine. However, we need to keep this clear for the users to minimize confusions. We need clear error messages so the user understands that assumptions are being violated.
That might be already done, but I have not had time to dive again into it.
Will try this week.

NLaws committed:
use PushVectors and Threads to speed up linearization

for example, with 4 threads check_upper_objective_for_bilinear_linearization goes from 28 seconds to less than 1 second with 10,000 variables

NLaws committed:
add @info for profiling src/bilinear_linearization.jl

NLaws committed:
add profiling @info to standard_form

NLaws committed:
speed up standard form for bilinear linearization

from ~35 minutes to less than 1 minute!

NLaws committed:
speed up standard form for bilinear linearization

from ~35 minutes to less than 1 minute!

NLaws committed:
merge bilinear branch into profiling branch

NLaws committed:
add GC.gc() to get_all_connected_rows_cols

NLaws committed:
add more profiling @info

NLaws committed:
Revert "add GC.gc() to get_all_connected_rows_cols"

This reverts commit 7c8cb7388d3c102e4c335706d3ac99b44f952c2c.

NLaws committed:
memoize find_connected_rows_cols

gets called with repeated arguments many times in the planning use case that I am developing

NLaws committed:
memoize find_connected_rows_cols

gets called with repeated arguments many times in the planning use case that I am developing

NLaws committed:
add GC.gc() to get_all_connected_rows_cols

NLaws committed:
add GC.gc() to get_all_connected_rows_cols

NLaws committed:
rm GC.gc() causing slow-down in get_all_connected_rows_cols

NLaws committed:
rm GC.gc() causing slow-down in get_all_connected_rows_cols

NLaws committed:
add generated recursive_col_search

@time memoized and threaded find_connected_rows_cols:
21.606794 seconds (2.95 M allocations: 166.152 MiB, 1.61% compilation time)

@time with generated recursive_col_search:
1.217451 seconds (1.41 M allocations: 74.890 MiB, 99.92% compilation time)

(tests done with first 8 values in upper_var_to_lower_ctr)

this commit is in case we need to go back to other methods and test times

next commit will (likely) switch to the generated method of recursive_col_search in find_connected_rows_cols (and use the globalV in more places perhaps)

NLaws committed:
implement generated code in bilinear linearization

hours of time turned into minutes (e.g. BilevelJuMP.get_all_connected_rows_cols went from 8.5 hours to 5.5 minutes)

NLaws committed:
fix mistake in recursive_col_search_expr

was not returning an expression and was returning too soon

NLaws committed:
rm redundant values in check_upper_objective_for_bilinear_linearization

the repeat function was repeating the same instance of vector to be filled, resulting in nthreads repeats in each vector (which causes slow downs in downstream code that loop over the outputs of this function)

NLaws committed:
use ThreadSafeDict to memoize find_connected_rows_cols

much faster using I,J,V from findnz in find_connected_rows_cols than prior speed up attempts. also, lower memory impact to use custom dict for memo cache because do not need to save the array input for each call of find_connected_rows_cols

NLaws committed:
thread another loop in linearization

NLaws committed:
Revert "thread another loop in linearization"

This reverts commit 70a3355b07570895e6d2e4ec8ac01b552fadfc03.

NLaws committed:
thread another loop in linearization

NLaws committed:
switch profiling info to @debug

NLaws committed:
switch @info to @debug in jump.jl

NLaws committed:
merge bilinear branch into profiling branch

NLaws committed:
Merge pull request #2 from NLaws/profiling

sped up some methods in linearization of bilinear terms by:
- memoizing find_connected_rows_cols
- multithreading several for loops
- working with lower level sparse coefficient matrix in its rows, cols, vals form (from findnz)

NLaws committed:
multi thread linearization terms for non empty AB

NLaws committed:
add check_linearization_conditions bool

can save time for large models if checks have already been done

NLaws committed:
add test/Project.toml JuMP compat =1.0

NLaws committed:
update test/bilinear_linearization.jl

new bool arg to main_linearization for checking conditions

NLaws committed:
add store_cache option to find_connected_rows_cols_cached

NLaws committed:
rm unnecessary argument in bilinear_linearization

NLaws committed:
update test/bilinear_linearization.jl

NLaws committed:
add some typing to biliinear_linearization

NLaws committed:
update linear_terms_for_non_empty_AB

- only need one pair of (j,n) from set A for each block in lower level problem
- see equ. 23 or 24 in https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=9729553

NLaws committed:
rm debug lines

joaquimg commented:
This is getting very nice.
Let me know whenever it is ready for another pass!

NLaws committed:
merge master

NLaws committed:
Pkg.rm("Gurobi")

was added for local tests

NLaws committed:
rm debug notes

NLaws commented:
@joaquimg ^^

NLaws commented:
@joaquimg ^^
